pytorch_lightning/callbacks/pruning.py

```diff
@@ -215,175 +217,183 @@ def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
         )
 
-    def apply_pruning(self, trainer: 'pl.Trainer', pl_module: LightningModule):
-        amount = self._resolve_amount(trainer.current_epoch)
-
-        # the user could control the pruning frequency with amount_fn
-        if amount == 0 or amount is None:
-            return
+    @staticmethod
+    def _get_pruned_stats(module: nn.Module, name: str) -> Tuple[int, int]:
+        attr = f"{name}_mask"
+        if not hasattr(module, attr):
+            return 0, 1
+        mask = getattr(module, attr)
+        return (mask == 0).sum().item(), mask.numel()
+
+    def apply_pruning(self, amount: Union[int, float]):
+        """ Applies pruning to ``parameters_to_prune``. """
+        if self._verbose:
+            stats = [self._get_pruned_stats(m, n) for m, n in self._parameters_to_prune]
 
         if self._use_global_unstructured:
             self._apply_global_pruning(amount)
         else:
             self._apply_local_pruning(amount)
 
-        if self._use_lottery_ticket_hypothesis:
-            self.apply_lottery_ticket_hypothesis()
+        if self._verbose:
+            for i, (module, name) in enumerate(self._parameters_to_prune):
+                prev_mask_zeroes, prev_mask_size = stats[i]
+                curr_mask_zeroes, curr_mask_size = self._get_pruned_stats(module, name)
+                rank_zero_info(
+                    f"Applied `{self.pruning_fn.__name__}` to `{module!r}.{name}` with amount={amount}. Pruned:"
+                    f" {prev_mask_zeroes} ({prev_mask_zeroes / prev_mask_size:.2%}) ->"
+                    f" {curr_mask_zeroes} ({curr_mask_zeroes / curr_mask_size:.2%})"
+                )
 
     def on_before_accelerator_backend_setup(self, trainer, pl_module):
         parameters_to_prune = self.sanitize_parameters_to_prune(
-            pl_module, self._parameters_to_prune, parameters=self._parameter_names
+            pl_module, self._parameters_to_prune, parameter_names=self._parameter_names
         )
 
         self._parameters_to_prune = self.filter_parameters_to_prune(parameters_to_prune)
 
         if self._use_lottery_ticket_hypothesis:
-            # make a copy of copy of orginal weights.
-            self._initial_parameters_to_prune = [(deepcopy(m), n) for m, n in self._parameters_to_prune]
-
-    def on_epoch_end(self, trainer, pl_module):
-        self.apply_pruning(trainer, pl_module)
-
-        if self.make_pruning_permanent:
-            self._make_pruning_permanent()
+            # group modules by id. Each entry has a copy of the initial data
+            # and a list of the associated parameter names to prune
+            self._original_layers = {}
+            for i, (module, name) in enumerate(self._parameters_to_prune):
+                id_ = id(module)
+                self._original_layers.setdefault(id_, {"data": deepcopy(module), "names": []})
+                self._original_layers[id_]["names"].append((i, name))
```

Comment:
Why not use self._original_layers[module] = name directly ?

Reply:
Because we wan't to remove duplicate modules.

For example:

```python
parameters_to_prune = [
    (a, "weight")
    (b, "weight")
    (a, "bias")
]
```

produces

```python
original_layers{
    id(a): {"data": deepcopy(a), "names": [(0, "weight"), (2, "bias)]},
    id(b): {"data": deepcopy(b), "names": [(1, "weight")]},
}
```

This avoids making duplicate copies of the modules while still having a reference to the original `parameters_to_prune` position. This reference is necessary because `_parameters_to_prune` is what's updated with each pruning